libgadget/sfr_eff.c

```diff
@@ -359,24 +489,29 @@ starformation(int i, double *localsfr, double * sum_sm, double * sum_mass_stars)
     double prob = P[i].Mass / mass_of_star * (1 - exp(-p));
 
     if(get_random_number(P[i].ID + 1) < prob) {
-        *sum_mass_stars += mass_of_star;
-        retflag = make_particle_star(i, mass_of_star);
+        /* ok, make a star */
+        newstar = i;
+        /* If we get a fraction of the mass we need to create
+         * a new particle for the star and remove mass from i.*/
+        if(P[i].Mass >= 1.1 * mass_of_star)
+            newstar = slots_split_particle(i, mass_of_star);
     }
 
-    if(P[i].Type == 0)	{
-        /* to protect using a particle that has been turned into a star */
+    /* A particle that will be turned into a star doesn't want the metals or to be made a wind.*/
+    if(newstar != i)	{
         SPHP(i).Metallicity += (1 - w) * METAL_YIELD * (1 - exp(-p));
 
         if(All.WindOn && HAS(All.WindModel, WIND_SUBGRID)) {
             /* Here comes the Springel Hernquist 03 wind model */
+            /*Notice that this is the mass of the gas particle after forking a star, 1/2 what it was before.*/
```

Comment:
Not necessarily half if GENERATIONS is not 2. It can be 1 / GENERATIONS

Reply:
Fixed